Collect error in the index

TristanCacqueray opened the pull request:
This PR enables collecting crawler errors so that the process does not stops when failing to process an entity.

TODO: add visual notification on the UI to indicate that some data may be missing when there are errors.

related: #1093

morucci commented on src/Monocle/Backend/Index.hs:
```diff
@@ -631,6 +649,17 @@ indexChanges changes = indexDocs $ fmap (toDoc . ensureType) changes
+  toDoc err = (getErrorDoc err, getErrorDocId err)
+
+  getErrorDoc :: EError -> Value
+  getErrorDoc err = object ["type" .= EErrorDoc, "error_data" .= toJSON err]
```
It seems that error_data should be base64 encoded: https://www.elastic.co/guide/en/elasticsearch/reference/current/binary.html

morucci replied:
More info about the binary type https://medium.com/@sachin.frayne/elasticsearch-binary-data-type-9429f8b2db85

morucci commented on src/Macroscope/Worker.hs:
```diff
@@ -105,12 +100,11 @@ processStream logFunc postFunc = go (0 :: Word) [] []
         pure $ reverse (res : results)
-      Right (Left err, rest) -> do
-        -- An error occured in the stream, abort now
-        let res = Just (StreamError (err, rest))
```
We have test that fail, I guess that's because now we commit the date. Should we update the test and make sure we get a Error doc instead ?

TristanCacqueray replied:
Yes, though we need to design the error query first.

TristanCacqueray commented:
Note that I haven't tested this change with real data.

TristanCacqueray commented on src/Lentille/GraphQL.hs:
```diff
@@ -254,15 +256,18 @@ streamFetch client@GraphClient {..} mkArgs StreamFetchOptParams {..} transformRe
         -- Yield the error and stop the stream
-        S.yield (Left $ GraphError e)
+        now <- lift mGetCurrentTime
+        S.yield (Left $ GraphError now e)
```
I guess here we should try to extract the pageInfo from the error to make the process continue, otherwise I'm not sure the current implementation will be able to resume after the error.

TristanCacqueray commented on src/Macroscope/Worker.hs:
```diff
@@ -105,12 +101,12 @@ processStream logFunc postFunc = go (0 :: Word) [] []
         -- We got a new document
+        let doc = case edoc of
+              Right x -> x
+              Left (DecodeError ts err) -> DTError $ CrawlerError "decode" (encodeBlob err) (Just $ from ts) (Just $ from entity)
```
We might want to perform `encodeBlob` in the document encoder by the api right before submitting to es.

TristanCacqueray commented on web/src/App.res:
```diff
@@ -290,6 +290,34 @@ module About = {
 }
 
+module Errors = {
+  module CrawlerError = {
```
This component could be improved to be more pretty :)

TristanCacqueray reviewed (commented):
@morucci the implementation is now mostly complete, but we might want to check it is working as expected, e.g. by trying to index the llvm project.

bigmadkev commented:
Just ran this pull request locally against the LLVM project I'm working with.
Can confirm that it posted 464 entries when it got stuck again.

Wonder if it should continue on through as it's not hit a no next page error? As a clean install will get stuck on this one.

TristanCacqueray commented:
@bigmadkev Thank you for testing this change. Could you please confirm that the crawler is actually stuck on the error? I would expect the "last update" timestamp to be set past the offending PR so that the crawler should be able to resume.

Though as indicated in the comment above, this implementation is presently skipping all the event happening after the error,  I'll propose a change to decode the pageinfo from the error when it is available.

bigmadkev commented:
I can see that the next run picks up all the updated items on the next run and doesn't hit the errored one again.

But the issue is that anything updated before the erroring one isn't indexed at all so rther than getting 9k it's only getting 500 ish up to the erroring pull request.

TristanCacqueray commented:
Thank you for confirming, so that's expected. It looks like from the crawler.log you shared that we are getting the [FetchErrorProducedErrors](https://hackage.haskell.org/package/morpheus-graphql-client-0.27.3/docs/Data-Morpheus-Client.html#t:FetchError) which actually contains the desired response, and I have updated this PR to handle that case. Make sure to delete your index if you want to try again. If you rebuild the web interface, you should see a red bell on the top right with a new errors page that should display `The additions count for this commit is unavailable.`

@morucci I am not sure this change is great because other unexpected errors might skip a large amount of data (because the crawler now always sets the last updated timestamp). Perhaps we should be more conservative in the `Worker.processStream` and only skip the PartialResult variant. In anycase, I have refactored the LentilleError and GraphQLError to help with further improvements.

Anyway I'll get back to this after the holidays, have a good end of the year! Cheers :)

morucci commented:
Well done ! Thanks !
